src/core/server/app/middleware/csp/tenant.ts

```diff
@@ -69,30 +72,56 @@ async function retrieveSiteFromEmbed(
   return null;
 }
 
-type Options = Pick<AppOptions, "mongo">;
+async function retrieveOriginsFromRequest(
+  mongo: Db,
+  req: Request
+): Promise<string[]> {
+  const site = await retrieveSiteFromEmbed(mongo, req);
+  if (!site) {
+    return [];
+  }
+
+  return site.allowedOrigins;
+}
+
+type Options = Pick<AppOptions, "mongo"> & {
+  /**
+   * frameAncestorsDeny when true will prevent the request's page from being
+   * embedded on any page inside an iframe.
+   */
+  frameAncestorsDeny?: true;
+};
 
 /**
  * cspMiddleware handles adding the CSP middleware to each outgoing request.
  */
-export const cspSiteMiddleware = ({ mongo }: Options): RequestHandler => async (
-  req,
-  res,
-  next
-) => {
-  // Initially, we do not allow any origins.
-  const site = await retrieveSiteFromEmbed(mongo, req);
+export const cspSiteMiddleware = ({
+  mongo,
+  frameAncestorsDeny,
+}: Options): RequestHandler => async (req, res, next) => {
+  res.setHeader("X-XSS-Protection", "1; mode=block");
 
-  // Grab the origins from the site.
-  const origins = site ? site.allowedOrigins : [];
+  // If the frame ancestors isn't being forced, and the tenant has AMP enabled,
+  // then don't add any headers!
+  if (
+    !frameAncestorsDeny &&
+    req.coral.tenant &&
+    hasFeatureFlag(req.coral.tenant, GQLFEATURE_FLAG.ENABLE_AMP)
+  ) {
+    return next();
+  }
```

Comment:
Isn't this going to let anyone embed Coral, when amp is enabled?

Reply:
Yes, though it will still require to use the permitted domains list for the story creation.